### PR TITLE
Tasklist filters change

### PR DESCRIPTION
**Pull request by golpeo**
Added 2 new components for the taklist filters, dropDownArrow and inputTagArrow that enclose dropdown and multiselect react-widgets.

The stack-client branch for this is named the same, comparison/diff with master:

https://github.com/macropodhq/stack-client/compare/master...tasklist-filters-change

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "macropod-components",
-  "version": "0.0.31",
+  "version": "0.0.32",
```
Comment (ojame) [inline package.json]: this is done in a separate commit (like `Bump version to v0.0.32`, add tag of `0.0.32`). in that commit you update the version in `package.json` and also update the changelog

File: packages/dropdown-arrow/dropdown-arrow.scss [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    position: fixed;
+    top: 0;
+    left: 0;
+    width: 100%;
+    height: 100%;
+    z-index: 1;
+  }
+  .rw-dropdownlist{
+    border: none;
+      position: absolute;
```
Comment (ojame) [inline packages/dropdown-arrow/dropdown-arrow.scss]: check the indentation in this selector

File: packages/dropdown-arrow/index.jsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    if (this.state.overlayShown) {
+      layer = (
+        <div onClick={this.changeHandler} className="overlay"></div>
+      );
+    }
+
+  	return (
+  		<div className="dropdown-arrow-container">
+  			<button onClick={this.changeHandler} className="trigger" >{this.props.label}<div/></button>
```
Comment (ojame) [inline packages/dropdown-arrow/index.jsx]: why does this need a stray `<div />`?

Comment (golpeo) [inline packages/dropdown-arrow/index.jsx]: For the arrows at the right in the label, done with css and pseudo selectors.

**Comment by ojame** [comment]
looks like there's some repeated styling across the two files - maybe you could dry it up a bit by centralising it? converting it to our `.mcss` and using `composes:` would be clean - not sure if you feel comfortable with that though.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,10 @@
 # Macropod Components Changelog
```
Comment (nelix) [inline CHANGELOG.md]: we should update this separately when we do the release bump

File: package.json (hunk 1 above)
Comment (golpeo) [inline package.json]: Done, changed package.json and changelog

File: packages/dropdown-arrow/dropdown-arrow.scss [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        background: #8d9daf;
+        color: #ffffff;
+      }
+    }
+  }
+}
+
+
+
+
```
Comment (nelix) [inline packages/dropdown-arrow/dropdown-arrow.scss]: why so many line feeds?

Comment (golpeo) [inline packages/dropdown-arrow/dropdown-arrow.scss]: No reason, fixed now.

File: packages/form/input-tag-arrow.jsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    var layer = null;
+
+    if (this.state.overlayShown) {
+      layer = (
+        <div onClick={this.changeHandler} className="overlay"></div>
+      );
+    }
+
+    var cx = React.addons.classSet;
```
Comment (ojame) [inline packages/form/input-tag-arrow.jsx]: depreciated. use `classname` instead. see: https://github.com/macropodhq/macropod-components/blob/master/packages/link-group/index.jsx#L6 for an example

**Comment by ojame** [comment]
make sure you use `let` in case of `var`, and `const` where required. there's a few `var`'s that snuck in :D

File: packages/form/input-tag-arrow.scss [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+          width: 10px;
+          top: 13px;
+          left: 7px;
+          content: '';
+          pointer-events: none;
+          background-color: #adb7c6;
+          height: 1px;
+          position: absolute;
+          -webkit-transform: rotate(-45deg);
+          -ms-transform: rotate(-45deg);
```
Comment (ojame) [inline packages/form/input-tag-arrow.scss]: we use autoprefixer, so vendors aren't required!

File: packages/form/input-tag-arrow.scss [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+      border-bottom-right-radius: 3px;
+      border-bottom-left-radius: 3px;
+      min-height: 28px;
+      input{
+        display: none;
+      }
+    }
+  }
+
+  .rw-select{
```
Comment (ojame) [inline packages/form/input-tag-arrow.scss]: emmmpty!

File: style/inputs.scss [hunk 9]
```diff
@@ -56,3 +47,69 @@ $input-state-error: #d46060;
...
+    top: 11px;
+    content: '';
+    pointer-events: none;
+    height: 1px;
+    position: absolute;
+    transform: rotate(-45deg);
+    background-color: #969798;
+    -webkit-backface-visibility: hidden;
+  }
+  div{
```
Comment (ojame) [inline style/inputs.scss]: i reckon we could merge a few of these styles to make it even dryer:

```
&:after, &:before {
   width: 5px;
   position: absolute;
   top: 16px;
   content: ' ';
   height: 1px;
   pointer-events: none;
   background-color: #969798;
   -webkit-backface-visibility: hidden;
}

&:after {
   right: 0;
   transform: rotate(-225deg);
}

&:before {
   right: 3px;
   transform: rotate(-135deg);
}
```

what are your thoughts on that?

File: packages/form/input-tag-arrow.scss (hunk 7 above)
Comment (golpeo) [inline packages/form/input-tag-arrow.scss]: fixed

File: packages/form/input-tag-arrow.scss (hunk 8 above)
Comment (golpeo) [inline packages/form/input-tag-arrow.scss]: removed

File: packages/form/input-tag-arrow.jsx (hunk 6 above)
Comment (golpeo) [inline packages/form/input-tag-arrow.jsx]: Using classnames now, cleaner.

File: CHANGELOG.md (hunk 4 above)
Comment (golpeo) [inline CHANGELOG.md]: Reverted version bump so its done by one of you.

File: style/inputs.scss (hunk 9 above)
Comment (golpeo) [inline style/inputs.scss]: right, i did the same for the inner div before/after styles

File: packages/dropdown-arrow/dropdown-arrow.scss [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  .rw-popup-container{
+    min-width: 120px;
+  }
+  .rw-popup{
+    border-color: #b2bdc9;
+    box-shadow: 1px 1px 2px rgba(0,0,0,.15);
+    border-top: 1px solid #b2bdc9;
+    border-top-left-radius: 3px;
+    border-top-right-radius: 3px;
+    ul{
```
Comment (ojame) [inline packages/dropdown-arrow/dropdown-arrow.scss]: space before `ul {` and none after

File: packages/dropdown-arrow/index.jsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+const Widgets = require('react-widgets');
+
+require('./dropdown-arrow.scss');
+
+module.exports = React.createClass({
+
+  displayName: 'DropDownArrow',
+
+  getInitialState() {
+    return { open: false, overlayShown: false  };
```
Comment (ojame) [inline packages/dropdown-arrow/index.jsx]: we like to format our objects on new lines:
```
return {
   open: false,
   overlayShown: false,
}
```

Make sure the last item has a trailing comma too

File: packages/dropdown-arrow/index.jsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  	this.setState({ open: !this.state.open, overlayShown: !this.state.open});
+  },
+
+  render() {
+
+    let layer = null;
+
+    if (this.state.overlayShown) {
+      layer = (
+        <div onClick={this.changeHandler} className="overlay"></div>
```
Comment (ojame) [inline packages/dropdown-arrow/index.jsx]: can self close:

`<div onClick={this.changeHandler} className="overlay" />`

File: packages/dropdown-arrow/index.jsx [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+      );
+    }
+
+  	return (
+  		<div className="dropdown-arrow-container">
+  			<button onClick={this.changeHandler} className="button-trigger" >{this.props.label}<div/></button>
```
Comment (ojame) [inline packages/dropdown-arrow/index.jsx]: watch the trailing space after `className="button-trigger" ` - don't need it.

File: packages/form/input-tag-arrow.jsx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+          open={this.state.open}
+          duration={50}
+        />
+      </div>
+    );
+  },
+});
+
+
+
```
Comment (ojame) [inline packages/form/input-tag-arrow.jsx]: only need one empty line at end of file

File: packages/form/input-tag-arrow.scss [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+      padding: 2px 21px 3px 7px;
+      font-size: 12px;
+      max-width: 95%;
+      .rw-btn{
+        top: 3px;
+      }
+    }
+
+  }
+}
```
Comment (ojame) [inline packages/form/input-tag-arrow.scss]: need new line at end of every file, it's how we roll :game_die:

File: packages/form/input-tag-arrow.jsx [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+      'multiselect-arrow-container':true
+    });
+
+    return (
+      <div className={classes}>
+        <button onClick={this.changeHandler} className="button-trigger">{this.props.label}<div/></button>
+        {layer}
+        <Widgets.Multiselect
+          {...this.props}
+          onClick={this.optionClickHandler}
```
Comment (ojame) [inline packages/form/input-tag-arrow.jsx]: where does this function come from?

File: packages/form/input-tag-arrow.jsx [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    let classes = cx({
+      'empty': this.state.isEmpty,
+      'multiselect-arrow-container':true
+    });
+
+    return (
+      <div className={classes}>
+        <button onClick={this.changeHandler} className="button-trigger">{this.props.label}<div/></button>
```
Comment (ojame) [inline packages/form/input-tag-arrow.jsx]: there should be a default prop for this

getDefaultProps() {

}

**Comment by ojame** [comment]
@golpeo we had a chat about implementing an example here, so we can see it. it's awesome to do this, because you've created two components, and if you can't preview them independently, it might be a bit of a pain. no point having an example which is coupled to stack, so:

this introduces two new components, `dropodown-arrow` and `input-tag-arrow` - variants of `dropdown` and `input-tag` respectively. Take the `input-tag` component:

- [input-tag](https://github.com/macropodhq/macropod-components/blob/master/packages/form/input-tag.jsx)
- [input-tag-arrow](https://github.com/macropodhq/macropod-components/blob/tasklist-filters-change/packages/form/input-tag-arrow.jsx)

They're both very similar both in code and output. I know one includes `InputWrapper` and the other doesn't, but i think you should **combine the variants**. You should have a result that can take an `arrow=true` prop to show the arrow, like this:

```
<InputTag
   arrow
/>
```

----

I've mashed together the two in a gist [here](https://gist.github.com/ojame/db9233f146bbe93511eb). Keep in mind this is **untested** and there might be an error or two somewhere. You should have an understanding of what this is trying to achieve though.

This way, in `macropod-components` you can have one example as what's already existing, and then you can have another example that shows the arrow, simply by passing the `arrow` prop.

What are your thoughts on that?

**Comment by golpeo** [comment]
Sounds like a good idea, at the end there will be 2 components enclosed in one but at least the component functionality is kept on one file/call, i'll merge and test the component and i'll use the same approach with the standard dropdown.

**Comment by golpeo** [comment]
Merged both components as per @ojame example, instead of 2 new components the same are kept with a extra `label` and `arrow` options, from the view standpoint everything remains the same.

File: packages/form/input-tag-arrow.jsx (hunk 16 above)
Comment (golpeo) [inline packages/form/input-tag-arrow.jsx]: Was an old one, removed with the new options

File: packages/form/input-tag-arrow.jsx (hunk 17 above)
Comment (golpeo) [inline packages/form/input-tag-arrow.jsx]: Fixed with new options added.

File: packages/dropdown/index.jsx [hunk 18]
```diff
@@ -9,7 +9,61 @@ require('./dropdown.scss');
...
+  inputWrapper() {
+    return (
+    	<Widgets.DropdownList {...this.props} />
+    );
+  },
+
+  arrowWrapper() {
+    let layer = null;
+
+    if (this.state.overlayShown) {
```
Comment (ojame) [inline packages/dropdown/index.jsx]: overlayShown needs to be in getInitialState, not getDefaultProps

File: packages/form/input-tag.jsx [hunk 19]
```diff
@@ -23,11 +59,39 @@ module.exports = React.createClass({
...
-        />
+        {this.multiSelect(normalisedProps)}
       </InputWrapper>
     );
   },
+
+  arrowWrapper() {
+    let layer = null;
+
+    if (this.state.overlayShown) {
```
Comment (ojame) [inline packages/form/input-tag.jsx]: needs to be in initial state

File: packages/form/input-tag.scss [hunk 20]
```diff
@@ -172,4 +172,114 @@
...
+        &:after {
+          width: 10px;
+          top: 13px;
+          left: 7px;
+          content: '';
+          pointer-events: none;
+          background-color: #adb7c6;
+          height: 1px;
+          position: absolute;
+          transform: rotate(-45deg);
```
Comment (ojame) [inline packages/form/input-tag.scss]: looks like some of the :before and :after can be combined (pointer-events, bg color, height, content, position)

File: packages/form/input-tag.jsx [hunk 21]
```diff
@@ -3,19 +3,55 @@
...
 
 module.exports = React.createClass({
+
   displayName: 'InputTag',
 
   getDefaultProps() {
     return {
       label: '',
+      arrow: false,
+      layerShown: false,
```
Comment (ojame) [inline packages/form/input-tag.jsx]: we should add some prop type validation, like: https://facebook.github.io/react/docs/reusable-components.html

Comment (golpeo) [inline packages/form/input-tag.jsx]: simple, done

File: packages/form/input-tag.jsx (hunk 19 above)
Comment (golpeo) [inline packages/form/input-tag.jsx]: also corrected in dropdown

File: packages/dropdown/index.jsx (hunk 18 above)
Comment (golpeo) [inline packages/dropdown/index.jsx]: corrected for both components

**Comment by ojame** [comment]
commits b221dc52c9b262a894b1d7d48215f77992caee6d and 9a0b087e9a9d000a4ccb822789087f75f849640a still refer to bumping version to 0.0.32 - this happened on master a while back so you should remove these two commits.

After that, probably squash a bunch of this into a couple of simple commits?

**Comment by ojame** [comment]
It also looks like the example still only shows the non-arrow version of the component. You should add the arrowed version of the component in there as well.

File: packages/dropdown/example.jsx [hunk 22]
```diff
@@ -38,7 +38,24 @@ module.exports = React.createClass({
...
+      />
+      <Dropdown
+        data={options}
+        value={this.state.selection}
+        onChange={this.handleChange}
+        textField="name"
+        label = "Options"
+        valueField="id"
+        duration={0}
+        arrow={true}
```
Comment (ojame) [inline packages/dropdown/example.jsx]: this can just be `arrow`, no need to have the `={true}`.

**Comment by ojame** [comment]
I'm getting a warning in the console:

```
WARNING in ./packages/dropdown/example.jsx
Module build failed: SyntaxError: /Users/jamescoleman/Documents/workspace/macropod-components/packages/dropdown/example.jsx: Adjacent JSX elements must be wrapped in an enclosing tag (49:6)
```
